rust/agents/relayer/src/msg/metadata/aggregation.rs

```diff
@@ -140,21 +147,34 @@ impl MetadataBuilder for AggregationIsmMetadataBuilder {
                 .map(|ism_address| self.base.build_ism(*ism_address)),
         )
         .await;
-
-        let filtered_sub_module_metas = metas
+        let (valid_sub_modules, invalid_sub_modules): (Vec<_>, Vec<_>) = metas
             .into_iter()
             .enumerate()
             .zip(sub_modules.into_iter())
-            .filter_map(|((index, meta_result), sub_module_result)| {
-                match (meta_result, sub_module_result) {
-                    (Ok(Some(meta)), Ok(ism)) => Some(IsmAndMetadata::new(ism, index, meta)),
-                    _ => None,
-                }
+            .partition(|&((_, ref meta_result), ref sub_module_result)| {
+                matches!((meta_result, sub_module_result), (Ok(Some(_)), Ok(_)))
+            });
+        let valid_sub_module_metas: Vec<_> = valid_sub_modules
+            .into_iter()
+            .map(|((index, meta_result), sub_module_result)| {
+                IsmAndMetadata::new(
+                    sub_module_result.unwrap(),
```

Comment:
We shouldn't have any unwraps, even if logically we know that `valid_sub_modules` when it's set above shouldn't panic here. It's a bit harder to understand and more prone to panics if we ever change code around here

If you take up my previous suggestion, we could do something like this. We have access to the itertools crate which gives us some extra power to return different left and right types using [partition_map](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.partition_map)

```
// Partitions things into
// ok_sub_modules: Vec<IsmAndMetadata>, which we'll then use later in a call to cheapest_valid_metas
// err_sub_modules: Vec<(H256 /* ism address */, Option<ModuleType>)>, which keeps the ISM address
//    even if there is an error getting the module type, and optionally has the module type if it was successful

let (ok_sub_modules, err_sub_modules): (Vec<IsmAndMetadata>, Vec<(H256 /* ism address */, Option<ModuleType>)>) = ism_and_metas
  .into_iter()
  .enumerate()
  .partition_map(|(index, (ism_address, ism_and_meta_result)) | {
    match (ism_and_meta_result) {
      // Everything successful
      Ok((ism, _module_type, Some(meta)) => Either::Left(IsmAndMetadata::new(ism, index, meta)),
      // Meta was None, but we still got module_type
      Ok((_ism, module_type, None) => Either::Right((ism_address, Some(module_type)),
      // Unable to get module_type, still keep the ism address
      _ => Either::Right((ism_address, None)),
    }
  });
```

Reply:
So, we're returning `(Box<dyn InterchainSecurityModule>, Option<Vec<u8>>, Module_Type>)` or `(Box<dyn InterchainSecurityModule>, Option<Vec<u8>>, Option<Module_Type>)`. If it's the latter, we'll be matching against different types in the `build` function instead of `build_ism_and_metadata`?

Reply:
`module_type()` revert can occur when we're calling build too?

Reply:
if this is the case, we don't need `Some(module_type)` in the partition, bc we already UnsupportedModuleType error?

Reply:
hmm reading this through rn

Reply:
See this comment https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2861#discussion_r1384867953